lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -3271,15 +3271,35 @@ bool SROA_Helper::DoScalarReplacement(GlobalVariable *GV,
   return true;
 }
 
-static void ReplaceConstantWithInst(Constant *C, Value *V,
+// Replaces uses of constant C in the current function
+// with V, when those uses are dominated by V.
+// Returns true if it was completely replaced.
+static bool ReplaceConstantWithInst(Constant *C, Value *V,
                                     IRBuilder<> &Builder) {
+  bool bReplacedAll = true;
   Function *F = Builder.GetInsertBlock()->getParent();
+  Instruction *VInst = dyn_cast<Instruction>(V);
+  // Lazily calculate dominance
+  DominatorTree DT;
+  bool Calculated = false;
+  auto Dominates = [&](llvm::Instruction *Def, llvm::Instruction *User) {
+    if (!Calculated) {
+      DT.recalculate(*F);
+      Calculated = true;
+    }
+    return DT.dominates(Def, User);
+  };
+
   for (auto it = C->user_begin(); it != C->user_end();) {
     User *U = *(it++);
     if (Instruction *I = dyn_cast<Instruction>(U)) {
       if (I->getParent()->getParent() != F)
         continue;
-      I->replaceUsesOfWith(C, V);
+      if (VInst && Dominates(VInst, I)) {
```

Comment:
This repeats a check performed here: https://github.com/microsoft/DirectXShaderCompiler/blob/1c66c4d798b02c62510a7b751303bea4cfd10849/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp#L3521

Can you explain why it's necessary? Maybe there's a recalculation missed somewhere? The intent of the original code was to minimize the number of times recalculation was needed. The new code here will recalculate it every time `ReplaceConstantWithInst` is called.

At the very least, the original check shouldn't be necessary

Reply:
The original check doesn't do the trick because the very first condition in `DominateAllUsers` is true:
```cpp
  // The Entry Block dominates everything, trivially true
  if (&F->getEntryBlock() == I->getParent())
    return true;
```
Even if we comment this out, and execute the code that computes the DT and calls `DominateAllUsersDom` below that, it still ends up returning true:
```cpp
  if (!DT) {
    DominatorTree TempDT;
    TempDT.recalculate(*F);
    return DominateAllUsersDom(I, V, &TempDT);
  } else {
    return DominateAllUsersDom(I, V, DT);
  }
```

The reason, I think, is because unlike in `ReplaceConstantWithInst`, `DominateAllUsersDom` doesn't handle the case of users of `V` being `ConstantExpr`. In the former, when the user is a `ConstantExpr`, it recurses with the `CE->getAsInstruction`, although it makes sure to insert it into the BB so that it's parent's parent is the current Function. I tried very hard to make `DominateAllUsersDom` do this, but it's tricky because in this case, we don't want to insert and instructions.

> At the very least, the original check shouldn't be necessary

Maybe? The thing is, this fix is only for the case of replacing a constant with a non-constant. I'm not sure if the other cases are properly avoided by this original check.

Reply:
I've done some additional experimentation and you are correct that the problem is replacing constants, which dominate everything, with non-constants, which don't unless you put them in the right place. At present, the new instructions are being placed after the memcpy, but that doesn't always dominate everything. It prevents the crash, but has some changes in the output of the pass. I'll get back to you with it shortly.

Reply:
Hi Greg, any update on this? Thanks!

Reply:
I've uploaded my incomplete take on this change in this draft PR #6591, we can discuss it there particularly if this causes compilation time regressions.